### PR TITLE
[REVIEW] Refactor some of the CSV Reader kernels into generic utility functions

### PR DESCRIPTION
**Pull request by vuule**
Set up utility functions for file parsing:
* Replace the countRecords kernel with thrust library calls;
* Make the launch_countRecords into a generic utility function;
* Make the storerecordStart kernel and launch_storerecordStart generic;
* move all of the above into a separate file, making the code reusable;
Other:
* Use Async CUDA calls for memory copies and sets;
* fix chunk allocation for small files - no longer allocates full chunk size when the file is smaller than chunk size;
* Refactor the code that set the record starts into a separate function;
* Rename the CSV specific typedefs to be more generic. usable in the utility functions;

### CODE REVIEW

**Comment by mjsamoht** [comment]
@OlivierNV @j-ieong can you help review this change?

File: cpp/src/io/utilities/parsing_utils.cu [hunk 1]
```diff
@@ -0,0 +1,143 @@
+#include "parsing_utils.cuh"
```
Comment (j-ieong) [inline cpp/src/io/utilities/parsing_utils.cu]: Missing copyright header

File: cpp/src/io/utilities/parsing_utils.cu [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+#include <cuda_runtime.h>
+
+#include <vector>
+
+#include "rmm/rmm.h"
+#include "rmm/thrust_rmm_allocator.h"
+#include "utilities/error_utils.hpp"
+
+constexpr size_t max_chunk_bytes = 256*1024*1024; // 256MB
```
Comment (j-ieong) [inline cpp/src/io/utilities/parsing_utils.cu]: There should be a comment on why this magic number was chosen.

File: cpp/src/io/utilities/parsing_utils.cu [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+	for (size_t ci = 0; ci < chunk_count; ++ci) {
+		const auto h_chunk = h_data + ci * max_chunk_bytes;
+		const auto chunk_bytes = std::min((size_t)(h_size - ci * max_chunk_bytes), max_chunk_bytes);
+
+		// Copy chunk to device
+		CUDA_TRY(cudaMemcpyAsync(d_chunk, h_chunk, chunk_bytes, cudaMemcpyDefault));
+
+		for (char key: keys) {
+			count += thrust::count(rmm::exec_policy()->on(0), d_chunk, d_chunk + chunk_bytes, key);
```
Comment (j-ieong) [inline cpp/src/io/utilities/parsing_utils.cu]: I'd rather we re-use the same kernel that we use for storing the start offsets. The logic for the paths to me is that we count in the first pass, then store in the second pass.

We re-use the same kernel in the Parquet reader; we launch to count the page headers, allocate memory, then launch to decode the page headers.

File: cpp/src/io/utilities/parsing_utils.cuh [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#pragma once
+
+#include <vector>
+
+#include "cudf.h"
+
+using ll_uint_t = unsigned long long int;
```
Comment (j-ieong) [inline cpp/src/io/utilities/parsing_utils.cuh]: Why the change in type name? I'd rather we keep the higher-level logical naming like pos_offset_type or similar (like gdf_size_type). Or we be explicit in type size like uint64_t.

File: cpp/src/io/utilities/parsing_utils.cuh [hunk 5]
```diff
@@ -0,0 +1,13 @@
+#pragma once
```
Comment (j-ieong) [inline cpp/src/io/utilities/parsing_utils.cuh]: Missing copyright header

File: cpp/src/io/utilities/parsing_utils.cu [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+		}
+	}
+
+	RMM_TRY( RMM_FREE( d_count, 0 ) ); 
+	RMM_TRY( RMM_FREE( d_chunk, 0 ) );
+
+	CUDA_TRY( cudaGetLastError() );
+
+	return GDF_SUCCESS;
+}
```
Comment (j-ieong) [inline cpp/src/io/utilities/parsing_utils.cu]: Missing newline at end of file. Some IDEs and tools require a line ending at the end

File: cpp/src/io/utilities/parsing_utils.cuh [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+#include "cudf.h"
+
+using ll_uint_t = unsigned long long int;
+
+gdf_error countAllFromSet(const char *h_data, size_t h_size, std::vector<char> keys, 
+	gdf_size_type &rec_cnt);
+
+gdf_error findAllFromSet(const char *h_data, size_t h_size, std::vector<char> keys, ll_uint_t result_offset,
+	ll_uint_t *recStart);
```
Comment (j-ieong) [inline cpp/src/io/utilities/parsing_utils.cuh]: Missing newline at end of file. Some IDEs and tools require a line ending at the end

File: cpp/src/io/utilities/parsing_utils.cu [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+		if (raw[i] == key) {
+			const auto pos = atomicAdd(count, 1ull);
+			positions[pos] = did + offset + i;
+		}
+	}
+}
+
+/**---------------------------------------------------------------------------*
+ * @brief For each of the characters in the input set, find and saves all 
+ * cccurences in the input host array of characters.
```
Comment (j-ieong) [inline cpp/src/io/utilities/parsing_utils.cu]: Typo in `cccurences`.

File: cpp/src/io/utilities/parsing_utils.cu [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+	RMM_TRY( RMM_FREE(d_chunk, 0) );
+
+	CUDA_TRY(cudaGetLastError());
+
+	return GDF_SUCCESS;
+}
+
+/**---------------------------------------------------------------------------*
+ * @brief CUDA kernel that finds all occurences of a character in the given
```
Comment (j-ieong) [inline cpp/src/io/utilities/parsing_utils.cu]: Typo in `occurences`

File: cpp/src/io/utilities/parsing_utils.cu [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+#include "rmm/rmm.h"
+#include "rmm/thrust_rmm_allocator.h"
+#include "utilities/error_utils.hpp"
+
+constexpr size_t max_chunk_bytes = 256*1024*1024; // 256MB
+constexpr int bytes_per_find_thread = 64;
+
+/**---------------------------------------------------------------------------*
+ * @brief Searches the input character array for each of characters in a set
+ * and sums up the number of occurences.
```
Comment (j-ieong) [inline cpp/src/io/utilities/parsing_utils.cu]: Typo in `occurences`

File: cpp/src/io/utilities/parsing_utils.cu [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+ *
+ * Does not load the entire buffer into the GPU memory at any time, so it can 
+ * be used with buffers of any size.
+ *
+ * @param[in] h_data Pointer to the data in host memory
+ * @param[in] h_size Size of the input data, in bytes
+ * @param[in] keys Vector containing the keys to count in the buffer
+ * @param[out] count Total number of occurences of all keys
```
Comment (j-ieong) [inline cpp/src/io/utilities/parsing_utils.cu]: Typo in `occurences`

File: cpp/src/io/utilities/parsing_utils.cuh (hunk 4 above)
Comment (vuule) [reply cpp/src/io/utilities/parsing_utils.cuh]: Used the alias because CUDA atomic operations are very picky about the parameter type. However, with the recent code refactoring, this is no longer needed. Changed to uint64_t.

File: cpp/src/io/utilities/parsing_utils.cu (hunk 3 above)
Comment (vuule) [reply cpp/src/io/utilities/parsing_utils.cu]: Done

**Comment by vuule** [comment]
@j-ieong Addressed all of your suggestions, please review.

Also refactored the code a bit, based on newly discovered JSON requirements, and the transition to exception-based error propagation (no return error code).
